### ISSUE TITLE
Autodetect free port just before starting myslqd

### ISSUE THREAD

**Issue by visadb** [vorburger/MariaDB4j#87]
Currently the port for MariaDB is autodetected way before mysqld actually reserves it, in `DBConfigurationBuilder.build()`. Eg. the mysld --bootstrap process is run between deciding the port and actually starting mysqld. That leaves ample time for the port to get reserved eg. on a busy CI server as in my case.

Should be easy to fix by deferring free port detection to `DB.start()`

**Comment by vorburger** [comment]
@visadb I see what you mean.. yeah, it probably indeed would be better if that port allocation was done lazily at the last moment (so not in the Builder but in the DB) - makes sense to me.

Would you be willing to contribute to this free project and raise a Pull Request which fixes this?

**Comment by visadb** [comment]
Sure, I'll send a PR.

**Comment by visadb** [comment]
Turns out that this wasn't so easy to fix. The default data dir contains the port number so that must be known when the bootstrap process is run. Also, the `.getURL()` and other methods depending on the port number would need to be moved to `DBConfiguration`.

So I ended up fixing the problem in my project by choosing the port myself and keeping the socket open until just before I call `DB.start()`. That is a bit ugly but is fine in the test code of my project.

I guess MariaDB4j could theoretically do the same thing: choose the port early but keep it reserved until starting myslqd. But that could lead to all kinds of issues in the general case so I guess it might not be a good idea to implement in MariaDB4j.
